lib/block-supports/border.php

```diff
@@ -103,105 +91,48 @@ function gutenberg_apply_border_support( $block_type, $block_attributes ) {
 			$border_width .= 'px';
 		}
 
-		$styles[] = sprintf( 'border-width: %s;', $border_width );
+		$border_block_styles['width'] = $border_width;
 	}
 
 	// Border color.
 	if (
 		$has_border_color_support &&
 		! gutenberg_should_skip_block_supports_serialization( $block_type, '__experimentalBorder', 'color' )
 	) {
-		$has_named_border_color  = array_key_exists( 'borderColor', $block_attributes );
-		$has_custom_border_color = isset( $block_attributes['style']['border']['color'] );
-
-		if ( $has_named_border_color || $has_custom_border_color ) {
-			$classes[] = 'has-border-color';
-		}
-
-		if ( $has_named_border_color ) {
-			$classes[] = sprintf( 'has-%s-border-color', $block_attributes['borderColor'] );
-		} elseif ( $has_custom_border_color ) {
-			$border_color = $block_attributes['style']['border']['color'];
-			$styles[]     = sprintf( 'border-color: %s;', $border_color );
-		}
+		$preset_border_color          = array_key_exists( 'borderColor', $block_attributes ) ? "var:preset|border-color|{$block_attributes['borderColor']}" : null;
+		$custom_border_color          = _wp_array_get( $block_attributes, array( 'style', 'border', 'color' ), null );
+		$border_block_styles['color'] = $preset_border_color ? $preset_border_color : $custom_border_color;
 	}
 
 	// Generate styles for individual border sides.
 	if ( $has_border_color_support || $has_border_width_support ) {
 		foreach ( $sides as $side ) {
-			$border = _wp_array_get( $block_attributes, array( 'style', 'border', $side ), false );
-
-			if ( is_array( $border ) && ! empty( $border ) ) {
-				$split_border_styles = gutenberg_generate_individual_border_classes_and_styles( $side, $border, $block_type );
-				$styles              = array_merge( $styles, $split_border_styles );
-			}
+			$border                       = _wp_array_get( $block_attributes, array( 'style', 'border', $side ), null );
+			$border_side_values           = array(
+				'width' => isset( $border['width'] ) && ! gutenberg_should_skip_block_supports_serialization( $block_type, '__experimentalBorder', 'width' ) ? $border['width'] : null,
+				'color' => isset( $border['color'] ) && ! gutenberg_should_skip_block_supports_serialization( $block_type, '__experimentalBorder', 'color' ) ? $border['color'] : null,
+				'style' => isset( $border['style'] ) && ! gutenberg_should_skip_block_supports_serialization( $block_type, '__experimentalBorder', 'style' ) ? $border['style'] : null,
+			);
```

Comment:
Could be abstracted.

There's a related discussion over at https://github.com/WordPress/gutenberg/pull/40332#issuecomment-1103230416 that questions just how much the style engine should know about. 

For example, consider whether we should pass theme.json `$merged_settings` as an argument. It would allow the style engine to test for skipped serialization.

The argument against is that it might create too many dependencies. We might get entangled in a spaghetti bowl of our own logic.  See: https://github.com/WordPress/gutenberg/pull/40332#issuecomment-1102450930

Reply:
Good question!

I think I'd lean toward keeping the idea of skipping serialization within the block support, largely for the dependency / spaghetti issue referred to in that comment. But one way to look at it, is that ideally if we can have the style engine support enough desired features (like potentially passing a selector to target a particular element for a block support — or including that in metadata), then for styles generated by the style engine, there might not be as much need to skip serialization (or we can reduce our reliance upon the feature).

So, for me, that points to keeping the skip serialization checks within the block support, rather than the style engine itself. But I don't feel too strongly about it 🙂

Reply:
Well put, thanks for articulating that. I agree.

Whether to "serialize" or not also seems a prima facie decision of block supports. The style engine should probably just deal with the raw data it's given for now. Better not to overthink our strategy before we jump out of the plane. 😄